packages/utilities/src/lib/sleep.ts

```diff
@@ -0,0 +1,77 @@
+export interface SleepOptions {
+	/**
+	 * When provided the corresponding `AbortController` can be used to cancel an asynchronous action.
+	 */
+	signal?: AbortSignal | undefined;
+
+	/**
+	 * Set to `false` to indicate that the scheduled `Timeout`
+	 * should not require the Node.js event loop to remain active.
+	 * @default true
+	 */
+	ref?: boolean | undefined;
+}
+
+export class AbortError extends Error {
+	public code: string;
+	public constructor(
+		message?: string,
+		options?: {
+			cause?: unknown;
+			code?: string;
+		}
+	) {
+		super(message, options);
+		this.name = 'AbortError';
+		this.code = options?.code ?? 'ERR_ABORT';
+	}
+}
+
+export const DOMException: typeof globalThis.DOMException =
+	globalThis.DOMException ??
+	(() => {
+		// DOMException was only made a global in Node v17.0.0,
+		// but we support >= v14
+		try {
+			atob('~');
+		} catch (err) {
+			return Object.getPrototypeOf(err).constructor;
+		}
+	})();
+
+/**
+ * Sleeps for the specified number of milliseconds.
+ * @param ms The number of milliseconds to sleep.
+ * @param value A value with which the promise is fulfilled.
+ * @see {@link sleepSync} for a synchronous version.
+ */
+export function sleep<T = undefined>(ms: number, value?: T, options?: SleepOptions): Promise<T> {
+	return new Promise((resolve, reject) => {
+		const timer: NodeJS.Timeout | number = setTimeout(() => resolve(value!), ms);
+		if (options?.signal) {
+			options.signal.addEventListener('abort', () => {
+				clearTimeout(timer);
+				reject(
+					new AbortError('The operation was aborted', {
+						cause: new DOMException('The operation was aborted', 'AbortError')
+					})
+				);
+			});
```

Comment:
We should listen handle the [`AbortSignal.reason`](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal/reason) property that's set when calling [`AbortController.abort(reason)`](https://developer.mozilla.org/en-US/docs/Web/API/AbortController/abort) and handle it accordingly.